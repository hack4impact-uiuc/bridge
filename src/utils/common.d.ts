declare module "@hack4impact-uiuc/bridge/common" {
  import { Props, ReactType } from "react";
  import {
    ColorProps,
    SpaceProps,
    LayoutProps,
    TypographyProps as StyledTypographyProps,
    BordersProps as StyledBordersProps,
    BoxShadowProps,
  } from "styled-system";

  export interface BaseProps extends Props<any> {
    as?: ReactType;
    className?: string;
  }
  export interface CommonProps
    extends BaseProps,
      ColorProps,
      SpaceProps,
      LayoutProps {}

  export interface TypographyProps extends BaseProps, StyledTypographyProps {}

  export interface BorderProps extends StyledBordersProps, BoxShadowProps {}
}
