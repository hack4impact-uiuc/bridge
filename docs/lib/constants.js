export const COMPONENTS_URL = '/components';
export const DESIGN_URL = '/design';
