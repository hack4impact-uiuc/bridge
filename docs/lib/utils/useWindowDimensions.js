import { useState, useEffect } from 'react';

function getWindowDimensions() {
  const { innerWidth: width, innerHeight: height } = window;
  return {
    width,
    height,
  };
}

export default function useWindowDimensions() {
  const [windowDimensions, setWindowDimensions] = useState(getWindowDimensions());

  useEffect(() => {
    function handleResize() {
      const { innerWidth: width, innerHeight: height } = window;
      setWindowDimensions({ width, height });
    }

    window.addEventListener('resize', handleResize);
    return () => window.removeEventListener('resize', handleResize);
  }, []);

  return windowDimensions;
}
