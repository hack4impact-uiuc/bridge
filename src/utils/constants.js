import * as styledSystem from 'styled-system';
import systemPropTypes from '@styled-system/prop-types';
import PropTypes from 'prop-types';


const { compose, system } = styledSystem;

// Text Decoration Prop
export const textDecoration = system({
  prop: 'textDecoration',
  cssProperty: 'text-decoration',
});
textDecoration.propTypes = PropTypes.string;


// common style props to be passed into components
export const COMMON = compose(styledSystem.space, styledSystem.layout);
COMMON.propTypes = {
  ...systemPropTypes.space,
  ...systemPropTypes.layout,
};

// for easier access
export const TYPOGRAPHY = styledSystem.typography;
export const LAYOUT = styledSystem.layout;
export const POSITION = styledSystem.position;
export const FLEX = styledSystem.flexbox;
export const GRID = styledSystem.grid;
export const COLOR = styledSystem.color;
export const BORDER = styledSystem.border;

TYPOGRAPHY.propTypes = systemPropTypes.typography;
LAYOUT.propTypes = systemPropTypes.layout;
POSITION.propTypes = systemPropTypes.position;
FLEX.propTypes = systemPropTypes.flexbox;
GRID.propTypes = systemPropTypes.grid;
COLOR.propTypes = systemPropTypes.color;
BORDER.propTypes = systemPropTypes.border;
