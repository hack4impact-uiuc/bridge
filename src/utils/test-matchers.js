/* eslint-disable react/forbid-foreign-prop-types */
// taken from https://github.com/primer/components/blob/master/src/utils/test-matchers.js
import 'jest-styled-components';

const stringify = (d) => JSON.stringify(d, null, '  ');
/**
 * These are props that styled-system aliases for backwards compatibility.
 * For some reason, they don't show up in our toImplementSystemProps() matcher,
 * so we skip over them.
 */
const ALIAS_PROP_TYPES = ['w', 'align', 'justify', 'wrap'];

// eslint-disable-next-line no-undef
expect.extend({
  toImplementSystemProps(Component, propNames) {
    const propKeys = new Set(Object.keys(Component.propTypes));
    const expectedPropKeys = Object.keys(propNames.propTypes);
    const missing = expectedPropKeys.filter((key) => !propKeys.has(key))
      .filter((key) => !ALIAS_PROP_TYPES.includes(key));
    return {
      pass: missing.length === 0,
      message: () => `Missing prop${missing.length === 1 ? '' : 's'}: ${stringify(missing)}`,
    };
  },
});
