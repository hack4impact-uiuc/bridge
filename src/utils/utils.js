import themeGet from '@styled-system/theme-get';
import * as styledSystem from 'styled-system';
import theme from '../theme';

const { get: getKey } = styledSystem;

export const get = (key) => themeGet(key, getKey(theme, key));

export const getOrDefault = (val, def) => (val !== undefined && val !== null ? val : def);

export const lodashGet = (object, path, defaultVal) => {
  const PATH = Array.isArray(path)
    ? path
    : path.split('.').filter((i) => i.length);
  if (!PATH.length) {
    return object === undefined ? defaultVal : object;
  }
  if (object === null || object === undefined || typeof (object[PATH[0]]) === 'undefined') {
    return defaultVal;
  }
  return lodashGet(object[PATH.shift()], PATH, defaultVal);
};
