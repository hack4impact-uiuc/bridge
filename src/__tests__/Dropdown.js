import React from 'react';
import { render, screen, cleanup } from '@testing-library/react';
import 'jest-styled-components';
import '@testing-library/jest-dom';
import { axe, toHaveNoViolations } from 'jest-axe';

import { Dropdown } from '..';
import { COMMON, TYPOGRAPHY } from '../utils/constants';
import { renderJSON } from '../utils/testing';
import 'babel-polyfill';
// axe violations required babel-polyfill
expect.extend(toHaveNoViolations);

describe('Dropdown', () => {
  it('should have no axe violations', async () => {
    const { container } = render(<Dropdown title="open" />);
    const results = await axe(container);
    expect(results).toHaveNoViolations();
    cleanup();
  });

  it('implements system props', () => {
    expect(Dropdown).toImplementSystemProps(COMMON);
    expect(Dropdown).toImplementSystemProps(TYPOGRAPHY);
  });

  it('renders correctly', () => {
    const item = renderJSON(
      <Dropdown>
        <Dropdown.Summary>Choose a color</Dropdown.Summary>
        <Dropdown.Menu>
          <Dropdown.Item>Item 1</Dropdown.Item>
          <Dropdown.Item>Item 2</Dropdown.Item>
          <Dropdown.Item>Item 3</Dropdown.Item>
        </Dropdown.Menu>
      </Dropdown>,
    );
    expect(item).toMatchSnapshot();
  });

  it('renders children', () => {
    const dropdownText = 'option';
    render(<Dropdown>{dropdownText}</Dropdown>);
    expect(screen.getByText(dropdownText)).toBeInTheDocument();
  });

  it('respects "width" props', () => {
    const { container } = render(<Dropdown width={200} />);
    expect(container.firstChild).toHaveStyle('width: 200px;');
  });

  it('respects "error" prop', () => {
    const item = renderJSON(<Dropdown error="error" />);
    expect(item).toMatchSnapshot();

    const item2 = renderJSON(<Dropdown error="success" />);
    expect(item2).toMatchSnapshot();

    const item3 = renderJSON(<Dropdown error="warning" />);
    expect(item3).toMatchSnapshot();
  });
});

describe('Dropdown.Item', () => {
  it('should have no axe violations', async () => {
    const { container } = render(<ul><Dropdown.Item /></ul>);
    const results = await axe(container);
    expect(results).toHaveNoViolations();
    cleanup();
  });

  it('renders children', () => {
    const dropdownText = 'option';
    render(<Dropdown.Item>{dropdownText}</Dropdown.Item>);
    expect(screen.getByText(dropdownText)).toBeInTheDocument();
  });

  it('implements system props', () => {
    expect(Dropdown.Item).toImplementSystemProps(COMMON);
    expect(Dropdown.Item).toImplementSystemProps(TYPOGRAPHY);
  });
});

describe('Dropdown.Summary', () => {
  it('should have no axe violations', async () => {
    const { container } = render(<Dropdown.Summary />);
    const results = await axe(container);
    expect(results).toHaveNoViolations();
    cleanup();
  });

  it('renders children', () => {
    const dropdownText = 'option';
    render(<Dropdown.Summary>{dropdownText}</Dropdown.Summary>);
    expect(screen.getByText(dropdownText)).toBeInTheDocument();
  });

  it('implements system props', () => {
    expect(Dropdown.Summary).toImplementSystemProps(COMMON);
    expect(Dropdown.Summary).toImplementSystemProps(TYPOGRAPHY);
  });
});

describe('Dropdown.Caret', () => {
  it('should have no axe violations', async () => {
    const { container } = render(<Dropdown.Caret />);
    const results = await axe(container);
    expect(results).toHaveNoViolations();
    cleanup();
  });

  it('implements system props', () => {
    expect(Dropdown.Caret).toImplementSystemProps(COMMON);
  });
});

describe('Dropdown.Menu', () => {
  it('should have no axe violations', async () => {
    const { container } = render(
      <Dropdown.Menu>
        <li key="a">1</li>
        <li key="b">2</li>
        <li key="c">3</li>
      </Dropdown.Menu>,
    );
    const results = await axe(container);
    expect(results).toHaveNoViolations();
    cleanup();
  });

  it('renders children', () => {
    const dropdownText = 'option';
    render(
      <Dropdown.Menu>
        {dropdownText}
      </Dropdown.Menu>,
    );
    expect(screen.getByText(dropdownText)).toBeInTheDocument();
  });

  it('implements system props', () => {
    expect(Dropdown.Menu).toImplementSystemProps(COMMON);
    expect(Dropdown.Menu).toImplementSystemProps(TYPOGRAPHY);
  });
});
