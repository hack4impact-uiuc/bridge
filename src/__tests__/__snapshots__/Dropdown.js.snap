// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown renders correctly 1`] = `
.c3 {
  background-clip: padding-box;
  background-color: #FFFFFF;
  border-radius: 2px;
  box-shadow: 0px 2px 4px rgba(0,0,0,0.15);
  left: 0;
  list-style: none;
  margin-top: 0px;
  padding: 0 !important;
  position: absolute;
  top: 100%;
  width: 100%;
  z-index: 100;
}

.c3::before {
  position: absolute;
  display: inline-block;
  content: '';
}

.c3::after {
  position: absolute;
  display: inline-block;
  content: '';
}

.c3 > ul {
  list-style: none;
}

.c3::before {
  top: -16px;
  left: 9px;
}

.c3::after {
  top: -14px;
  left: 10px;
}

.c2 {
  display: inline-block;
  float: right;
  margin-left: 12px;
  margin-bottom: 1px;
  vertical-align: middle;
}

.c4 {
  display: block;
  padding: 12px 10px 12px 14px;
  overflow: hidden;
  color: #272A2D;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.c4 a {
  color: #272A2D;
  -webkit-text-decoration: none;
  text-decoration: none;
  display: block;
  padding: 12px 10px 12px 14px;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.c4:focus,
.c4 a:focus {
  -webkit-text-decoration: none;
  text-decoration: none;
  background-color: rgba(21,93,161,0.05);
}

.c4:hover,
.c4:hover a {
  -webkit-text-decoration: none;
  text-decoration: none;
  background-color: rgba(21,93,161,0.05);
  outline: none;
}

.c1 {
  background: #FFFFFF;
  color: #8B9199;
  border-radius: 4px;
  padding: 12px;
  pointer: cursor;
}

.c0 {
  font-family: HKGrotesk;
  font-weight: 400px;
  font-size: 18px;
  -webkit-letter-spacing: 0.3px;
  -moz-letter-spacing: 0.3px;
  -ms-letter-spacing: 0.3px;
  letter-spacing: 0.3px;
  line-height: 20px;
  position: relative;
  display: inline-block;
  pointer: cursor;
  width: 100%;
}

.c0 > summary {
  list-style: none;
  border: 1px solid #8B9199;
  outline: none;
}

.c0[open] > summary {
  border: 2px solid #155DA1;
}

.c0 > summary::-webkit-details-marker {
  display: none;
}

<details
  className="c0"
  onToggle={[Function]}
  open={false}
>
  <summary
    aria-haspopup="true"
    className="c1"
  >
    Choose a color
    <span
      className="c2"
    >
      <svg
        fill="none"
        height="10"
        viewBox="0 0 14 8"
        width="14"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M1 1L7 7L13 1"
          stroke="#272A2D"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth="2"
        />
      </svg>
    </span>
  </summary>
  <ul
    className="c3"
  >
    <li
      className="c4"
    >
      Item 1
    </li>
    <li
      className="c4"
    >
      Item 2
    </li>
    <li
      className="c4"
    >
      Item 3
    </li>
  </ul>
</details>
`;

exports[`Dropdown respects "error" prop 1`] = `
.c0 {
  font-family: HKGrotesk;
  font-weight: 400px;
  font-size: 18px;
  -webkit-letter-spacing: 0.3px;
  -moz-letter-spacing: 0.3px;
  -ms-letter-spacing: 0.3px;
  letter-spacing: 0.3px;
  line-height: 20px;
  position: relative;
  display: inline-block;
  pointer: cursor;
  width: 100%;
}

.c0 > summary {
  list-style: none;
  border: 1px solid #A62B31;
  outline: none;
}

.c0[open] > summary {
  border: 2px solid #A62B31;
}

.c0 > summary::-webkit-details-marker {
  display: none;
}

<details
  className="c0"
  onToggle={[Function]}
  open={false}
/>
`;

exports[`Dropdown respects "error" prop 2`] = `
.c0 {
  font-family: HKGrotesk;
  font-weight: 400px;
  font-size: 18px;
  -webkit-letter-spacing: 0.3px;
  -moz-letter-spacing: 0.3px;
  -ms-letter-spacing: 0.3px;
  letter-spacing: 0.3px;
  line-height: 20px;
  position: relative;
  display: inline-block;
  pointer: cursor;
  width: 100%;
}

.c0 > summary {
  list-style: none;
  border: 1px solid #4C9859;
  outline: none;
}

.c0[open] > summary {
  border: 2px solid #4C9859;
}

.c0 > summary::-webkit-details-marker {
  display: none;
}

<details
  className="c0"
  onToggle={[Function]}
  open={false}
/>
`;

exports[`Dropdown respects "error" prop 3`] = `
.c0 {
  font-family: HKGrotesk;
  font-weight: 400px;
  font-size: 18px;
  -webkit-letter-spacing: 0.3px;
  -moz-letter-spacing: 0.3px;
  -ms-letter-spacing: 0.3px;
  letter-spacing: 0.3px;
  line-height: 20px;
  position: relative;
  display: inline-block;
  pointer: cursor;
  width: 100%;
}

.c0 > summary {
  list-style: none;
  border: 1px solid #FDBE21;
  outline: none;
}

.c0[open] > summary {
  border: 2px solid #FDBE21;
}

.c0 > summary::-webkit-details-marker {
  display: none;
}

<details
  className="c0"
  onToggle={[Function]}
  open={false}
/>
`;
